crates/runtime/src/instance/allocator.rs

```diff
@@ -104,104 +282,160 @@ pub unsafe trait InstanceAllocator {
+    /// Allocate the memories for the given instance allocation request, pushing
+    /// them into `memories`.
+    ///
+    /// # Unsafety
     ///
-    /// The return value here, if successful, is passed to the various methods
-    /// below for memory/table allocation/deallocation.
-    fn allocate_index(&self, req: &InstanceAllocationRequest) -> Result<usize>;
+    /// The request's associated module and memories must have previously been
+    /// validated by `Self::validate_module`.
+    unsafe fn allocate_memories(
+        &self,
+        request: &mut InstanceAllocationRequest,
+        memories: &mut PrimaryMap<DefinedMemoryIndex, (MemoryAllocationIndex, Memory)>,
+    ) -> Result<()> {
+        let module = request.runtime_info.module();
+
+        #[cfg(debug_assertions)]
+        InstanceAllocatorImpl::validate_module_impl(self, module, request.runtime_info.offsets())
+            .expect("module should have already been validated before allocation");
+
+        for (memory_index, memory_plan) in module
+            .memory_plans
+            .iter()
+            .skip(module.num_imported_memories)
+        {
+            let memory_index = module
+                .defined_memory_index(memory_index)
+                .expect("should be a defined memory since we skipped imported ones");
+
+            memories.push(self.allocate_memory(request, memory_plan, memory_index)?);
+        }
 
-    /// Deallocates indices allocated by `allocate_index`.
-    fn deallocate_index(&self, index: usize);
+        Ok(())
+    }
 
-    /// Attempts to allocate all defined linear memories for a module.
+    /// Deallocate all the memories in the given primary map.
     ///
-    /// Pushes all memories for `req` onto the `mems` storage provided which is
-    /// already appropriately allocated to contain all memories.
+    /// # Unsafety
     ///
-    /// Note that this is allowed to fail. Failure can additionally happen after
-    /// some memories have already been successfully allocated. All memories
-    /// pushed onto `mem` are guaranteed to one day make their way to
-    /// `deallocate_memories`.
-    fn allocate_memories(
+    /// The memories must have previously been allocated by
+    /// `Self::allocate_memories`.
+    unsafe fn deallocate_memories(
         &self,
-        index: usize,
-        req: &mut InstanceAllocationRequest,
-        mems: &mut PrimaryMap<DefinedMemoryIndex, Memory>,
-    ) -> Result<()>;
-
-    /// Deallocates all memories provided, optionally reclaiming resources for
-    /// the pooling allocator for example.
-    fn deallocate_memories(&self, index: usize, mems: &mut PrimaryMap<DefinedMemoryIndex, Memory>);
+        memories: &mut PrimaryMap<DefinedMemoryIndex, (MemoryAllocationIndex, Memory)>,
+    ) {
+        for (memory_index, (allocation_index, memory)) in mem::take(memories) {
+            self.deallocate_memory(memory_index, allocation_index, memory);
```

Comment:
request for comment: deallocation needs to be infallible, because if it wasn't we could leak if say the first of two memories failed to drop.